### PR TITLE
Introduce a better system for keeping track if a buffer is visible or not in tagging

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
IDocumentTrackingService is not great (it doesn't work in repl i've found).  This is a much clearer and cleaner association with the textview itself and if it's visible.

This is extracted from teh Pause-Taggers PR here https://github.com/dotnet/roslyn/pull/60371.  That way Pause/Unpause is in an independent chunk we merge.

### CODE REVIEW

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs [hunk 1]
```diff
@@ -563,6 +563,7 @@ private Task ProduceTagsAsync(TaggerContext<TTag> context, CancellationToken can
             public IEnumerable<ITagSpan<TTag>> GetTags(NormalizedSnapshotSpanCollection requestedSpans)
             {
                 this.AssertIsForeground();
+
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs:566]:
```suggestion
```

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs [hunk 2]
```diff
@@ -267,14 +267,13 @@ private void RaiseDiagnosticsRemovedIfRequiredForClosedOrResetDocument(TextDocum
...
-            {
-                return;
-            }
+            // Retrigger analysis of newly active document to always get up-to-date diagnostics.
+            // Note that we do so regardless of the current background analysis scope,
+            // as we might have switched the document _while_ the diagnostic refresh was in progress for
+            // all open documents, which can lead to cancellation of diagnostic recomputation task
+            // for the newly active document.  This can lead to a race condition where we end up with
+            // stale diagnostics for the active document.  We avoid that by always recomputing
+            // the diagnostics for the newly active document whenever active document is switched.
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs:276]: this is a merge of https://github.com/dotnet/roslyn/pull/60365

File: src/EditorFeatures/Core.Wpf/Workspaces/WpfTextBufferVisibilityTracker.cs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+        public bool IsVisible(ITextBuffer subjectBuffer)
+        {
+            Contract.ThrowIfFalse(_threadingContext.HasMainThread);
+
+            var views = _associatedViewService.GetAssociatedTextViews(subjectBuffer).ToImmutableArrayOrEmpty();
+
+            // if any of the views were *not* wpf text views, assume the buffer is visible.  We don't know how to
+            // determine the visibility of this buffer.
+            if (views.Any(v => v is not IWpfTextView))
```
Comment (dibarbet) [inline src/EditorFeatures/Core.Wpf/Workspaces/WpfTextBufferVisibilityTracker.cs:71]: what would be a valid case of us using this and having a non wpf view?  Is expected that calling this for a C#/vb/ts/f# buffer would not be?

Just wondering if this a bug in the caller and should throw

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs [hunk 4]
```diff
@@ -79,8 +79,9 @@ private sealed partial class TagSource : ForegroundThreadAffinitizedObject
 
             #region Fields that can only be accessed from the foreground thread
 
-            private readonly ITextView _textViewOpt;
+            private readonly ITextView? _textView;
             private readonly ITextBuffer _subjectBuffer;
+            private readonly ITextBufferVisibilityTracker? _visibilityTracker;
```
Comment (dibarbet) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs:84]: i presume the actual usage of this is in the followup pr

File: src/EditorFeatures/Core.Wpf/Workspaces/WpfTextBufferVisibilityTracker.cs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+            public void UpdateAssociatedViews()
+            {
+                Contract.ThrowIfFalse(_tracker._threadingContext.HasMainThread);
+
+                // Update us to whatever the currently associated text views are for this buffer.
+                UpdateTextViews(_tracker._associatedViewService.GetAssociatedTextViews(_subjectBuffer));
+            }
+
+            private void UpdateTextViews(IEnumerable<ITextView> associatedTextViews)
```
Comment (dibarbet) [inline src/EditorFeatures/Core.Wpf/Workspaces/WpfTextBufferVisibilityTracker.cs:144]: this updates the associated views (aka subscribes / unsubscribes to IsVisibleChanged).  But when this is called from either registering a text buffer or subject buffer connection/disconnection, should that also fire a call to the callback?
Or does each of those implicitly trigger `VisualElement.IsVisibleChanged` to fire in the editor?

File: src/EditorFeatures/Core.Wpf/Workspaces/WpfTextBufferVisibilityTracker.cs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+            _associatedViewService = associatedViewService;
+            _threadingContext = threadingContext;
+
+            associatedViewService.SubjectBuffersConnected += AssociatedViewService_SubjectBuffersConnected;
+            associatedViewService.SubjectBuffersDisconnected += AssociatedViewService_SubjectBuffersDisconnected;
+        }
+
+        private void AssociatedViewService_SubjectBuffersConnected(object sender, SubjectBuffersConnectedEventArgs e)
+        {
+            Contract.ThrowIfFalse(_threadingContext.HasMainThread);
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Workspaces/WpfTextBufferVisibilityTracker.cs:42]: FTAO is affectively just a legacy wrapper around IThreadingContext.  I'd prefer we just remove FTAO and use IThreadingContext in all places instead :)

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs:84]: Ooops! this was meant to be used.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs:84]: Ok.  Updated to use this.

File: src/EditorFeatures/Core.Wpf/Workspaces/WpfTextBufferVisibilityTracker.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/Workspaces/WpfTextBufferVisibilityTracker.cs:144]: I'll add docs.  But hte general idea is that the consumer of this should check the visibility of the buffer (using IsVisible) and then attach the listeners.  as this is all UI thread affinitized, this should always be accurate.

File: src/EditorFeatures/Core/Workspaces/ITextBufferVisibilityTracker.cs [hunk 7]
```diff
@@ -28,4 +32,57 @@ internal interface ITextBufferVisibilityChangedCallback
...
+
+            await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            if (service.IsVisible(subjectBuffer))
+                return;
+
+            // ensure we listen for visibility changes before checking.  That way we don't have a race where we check
+            // something see it is not visible, but then do not hear about its visibility change because we've hooked up
+            // our event after that happens.
+            var visibilityChangedTaskSource = new TaskCompletionSource<bool>();
+            var callback = new VisibilityCallback(visibilityChangedTaskSource);
```
Comment (dibarbet) [inline src/EditorFeatures/Core/Workspaces/ITextBufferVisibilityTracker.cs:68]: so this brings me to a question on the original API - any reason why the callback isn't a func or something similar?  Seems like each consumer needs a bit more boilerplate to use the ITextBufferVisibilityChangedCallback

File: src/EditorFeatures/Core/Workspaces/ITextBufferVisibilityTracker.cs [hunk 8]
```diff
@@ -22,9 +22,16 @@ internal interface ITextBufferVisibilityTracker
         bool IsVisible(ITextBuffer subjectBuffer);
 
         /// <summary>
-        /// Registers to hear about visibility changes for this particular buffer.
+        /// Registers to hear about visibility changes for this particular buffer.  Not: registration will not trigger a
```
Comment (dibarbet) [inline src/EditorFeatures/Core/Workspaces/ITextBufferVisibilityTracker.cs:25]: nit - Note:

**Review by dibarbet — APPROVED** [review]

**Comment by allisonchou** [comment]
Re-targeting branch from release/dev17.3 to main-vs-deps so we can delete the release/dev17.3 branch.

**Pull request merged**